lib/docker_tag_pusher.rb

```diff
@@ -0,0 +1,73 @@
+require 'json'
+require 'net/http'
+require 'uri'
+
+class DockerTagPusher
+  REGISTRY = 'https://registry-1.docker.io'.freeze
+  MEDIA_TYPE = 'application/vnd.docker.distribution.manifest.v2+json'.freeze
+
+  attr_reader :username, :password
+  private :username, :password
+
+  def initialize(username, password)
+    @username = username
+    @password = password
+  end
+
+  def get_manifest(repo, tag)
+    request = Net::HTTP::Get.new(
+      "#{REGISTRY}/v2/#{repo}/manifests/#{tag}",
+      'Authorization' => "Bearer #{token(repo)}",
+      'Accept' => MEDIA_TYPE
+    )
+    response = registry_client.request(request)
+
+    if response.code == '404'
+      raise 'Image or tag not found'
+    elsif response['Content-Type'] != MEDIA_TYPE
+      raise "Remote image not in correct format (must be #{MEDIA_TYPE})"
+    elsif response.code != '200'
+      raise "Server error while fetching manifest: #{response.body}"
+    else
+      response.body
+    end
```

Comment:
Since each of the `raise` conditionals will terminate the conditional you can probably simplify this down to a few guard clauses. eg:

```
raise 'Image or tag not found' if response.code == '404'
...
response.body
```